src/backend/vulkan/src/lib.rs

```diff
@@ -1216,7 +1216,7 @@ impl queue::CommandQueue<Backend> for CommandQueue {
         let submit_info = vk::SubmitInfo {
```

Comment:
Why is this queue submission necessary? Could we pass `wait_semaphore` to `vk::PresentInfoKHR` directly?

Reply:
If there is a wait semaphore, we should skip the submission. If there isn't, we need *a* semaphore to pass to the present(), so this is what the submission is used for.
Would you mind trying this change out?

Reply:
I see. In the case there is no wait semaphore, we must wait for the `COLOR_ATTACHMENT_OUTPUT` pipeline stage to complete, and we cannot use pipeline barriers to do so, because pipeline barriers exist in the context of a command buffer—and this is why we require a semaphore, because we must use a submission to wait for the pipeline stage?

Reply:
The idea is that an image of the new swapchain model is only going to be used in a render pass. User is still responsible to set up the initial/final layouts so that the image is presentable after it's rendered to. I.e. the quad example has `i::Layout::Undefined .. i::Layout::Present`.

We need a semaphore to synchronize rendering with presentation because Vulkan spec requires that. It allows the driver to treat presentation mechanism as an internal "queue", so it needs to be synchronized with the rendering queue.